crypto/bio/bio_lib.c

```diff
@@ -242,6 +242,19 @@ int BIO_method_type(const BIO *b)
     return b->method->type;
 }
 
+int BIO_method_type_final(const BIO *b)
+{
+    BIO *lb;
+
+    if (b == NULL)
+        return 0;
+    lb = b;
+    while (lb->next_bio != NULL)
+        lb = lb->next_bio;
```

Comment:
This seems to be too general, what if the intermediate FD is some kind
of tunnelling protocol?

Reply:
I don't understand why should this matter?

AFAIU, this should work fine as long as the final destination is a socket that uses the record layer to protect data. All intermediate BIOs will perform their duties correctly, and the final socket/fd will use KTLS for session protection.

Reply:
Hey, and what if the intermediate BIOs want to see the encrypted data?

Reply:
Is there any specific use-case you are referring to?

I think that such use cases should disable KTLS explicitly, i.e. use SSL_MODE_NO_KTLS_RX/SSL_MODE_NO_KTLS_TX.

Reply:
Sorry, _you_ should come up with a use case and a test case that
requires looking through an arbitrary stack of BIO filters.

Reply:
Should I invent a new BIO type that looks into encrypted payload only so that we test this hypothetical flow?

Reply:
No, but this PR definitely needs a test case.
You try to fix something, therefore you must have some use case in mind.
I am just telling you that your fix should probably be more specific.

Reply:
The test already exists.
This fixes KTLS tests, which are not passing in the master at the moment (see [PR9648](https://github.com/openssl/openssl/pull/9648))

Reply:
If you are using KTLS then it seems to me that there is no encrypted data in userspace; that's the whole point.